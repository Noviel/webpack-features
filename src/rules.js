const isObject = val => typeof val === 'object' && val !== null;

export function createLoader(
  name, 
  options = {}, 
  { nameOverride = null } = {}
) {
  if (typeof name !== 'string') {
    throw new Error('name should be a string');
  }

  let loader = `${name}`;
  
  if (typeof nameOverride === 'string' ) {
    loader = nameOverride;
  } else if (typeof nameOverride === 'function') {
    loader = nameOverride(name, options);
  } 

  return { 
    loader,
    options: { ...options } 
  };
}

export function createRule(loaders, { 
  exclude = /node_modules/,
  ...restOptions
} = {}) {
  if (!isObject(loaders)) {
    throw new Error('loaders should be an object');
  }

  return {
    use: loaders,
    exclude,
    ...restOptions
  };
}
