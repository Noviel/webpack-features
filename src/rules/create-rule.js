const isObject = val => typeof val === 'object' && val !== null;

export default function createRule(loaders, { 
  exclude = /node_modules/,
  ...restOptions
} = {}) {
  if (!isObject(loaders)) {
    throw new Error('loaders should be an object');
  }

  return {
    use: loaders,
    exclude,
    ...restOptions
  };
}
