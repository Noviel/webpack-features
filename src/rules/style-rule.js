import { createCSSLoader } from '../loaders/style-chain-loader';

export const createStyleRule = ({ target, production }, { 
  preprocessor,
  postcss, 
  modules = false, 
  modulesPath 
}, restOptions = {}) => {
  const loader = createCSSLoader(
    { target, production, useStyleLoader: true },
    { modules, minimize: production },
    (props, { target, production }) => {
      return {
        ...props,
        options: {
          ...props.options,
          localIdentName: production ? '[hash:base64]' : '[path][name]__[local]--[hash:base64:5]' 
        }
      };
    }
  );

  if (postcss) {
    loader.add('postcss-loader', postcss);
  }

  if (preprocessor === 'scss') {
    loader.add('sass-loader');
  } else if (preprocessor === 'less') {
    loader.add('less-loader');
  }

  const test =  preprocessor === 'scss' ? /\.scss$/i :
                preprocessor === 'less' ? /\.less$/i :
                preprocessor === 'emotion' ? /emotion\.css$/ :
                /\.css$/i;

  const options = {};

  if (modules) {
    options.include = modulesPath;
    options.exclude = /emotion\.css$/;
  } else {
    if (preprocessor !== 'emotion') {
      options.exclude = modulesPath;
    }
  }
  
  return {
    test,
    use: loader.get(),
    ...options,
    ...restOptions
  };
};
