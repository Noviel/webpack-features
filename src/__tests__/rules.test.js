import { createLoader, createRule } from '../rules';

describe('createLoader', () => {
  it('should throw with undefined name', () => {
    expect(() => createLoader())
      .toThrow();
  });

  it('should throw with non-string name', () => {
    expect(() => createLoader(420))
      .toThrow();
  });

  it('should override name with a string', () => {
    expect(createLoader('name', {}, { nameOverride: 'override' }))
      .toMatchSnapshot();
  });

  it('should override name with a function', () => {
    expect(createLoader('name', {}, { 
      nameOverride: name => `${name}-override`
    }))
      .toMatchSnapshot();
  });

  it('should return correct loader name', () => {
    expect(createLoader('strange-name-2').loader)
      .toMatchSnapshot();
  });

  it('should return correct options', () => {
    expect(createLoader('name', { a: { option: true }, b: 'option' }))
      .toMatchSnapshot();
  });
});

describe('createRule', () => {
  it('should throw with undefined loaders', () => {
    expect(() => createRule())
      .toThrow();
  });

  it('should assign default exclude value', () => {
    expect(createRule({ loader: 'loader' }))
      .toMatchSnapshot();
  });

  it('should merge provided options', () => {
    expect(
      createRule(
        { 
          loader: 'simple' 
        }, 
        {
          test: /\.jsx?$/
        }
      )
    )
      .toMatchSnapshot();
  });
});
