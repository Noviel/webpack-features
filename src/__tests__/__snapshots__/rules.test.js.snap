// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createLoader should override name with a function 1`] = `
Object {
  "loader": "name-override",
  "options": Object {},
}
`;

exports[`createLoader should override name with a string 1`] = `
Object {
  "loader": "override",
  "options": Object {},
}
`;

exports[`createLoader should return correct loader name 1`] = `"strange-name-2"`;

exports[`createLoader should return correct options 1`] = `
Object {
  "loader": "name",
  "options": Object {
    "a": Object {
      "option": true,
    },
    "b": "option",
  },
}
`;

exports[`createRule should assign default exclude value 1`] = `
Object {
  "exclude": /node_modules/,
  "use": Object {
    "loader": "loader",
  },
}
`;

exports[`createRule should merge provided options 1`] = `
Object {
  "exclude": /node_modules/,
  "test": /\\\\\\.jsx\\?\\$/,
  "use": Object {
    "loader": "simple",
  },
}
`;
